scene/gui/foldable_container.cpp

```diff
@@ -0,0 +1,301 @@
+	flags.append(SIZE_SHRINK_END);
+	return flags;
+}
+
+void FoldableContainer::set_expanded(bool p_expanded) {
+	expanded = p_expanded;
+	for (int i = 0; i < get_child_count(); i++) {
+		Control *c = Object::cast_to<Control>(get_child(i));
+		if (!c || c->is_set_as_top_level()) {
+			continue;
+		}
+		c->set_visible(expanded);
+	}
+	queue_redraw();
+}
+
+bool FoldableContainer::is_expanded() const {
+	return expanded;
+}
+
+void FoldableContainer::set_title(String p_title) {
+	if (title == p_title) {
+		return;
+	}
+	title = p_title;
+	queue_redraw();
+}
+
+String FoldableContainer::get_title() const {
+	return title;
+}
+
+void FoldableContainer::set_title_alignment(HorizontalAlignment p_alignment) {
+	ERR_FAIL_INDEX((int)p_alignment, 3);
+	if (title_alignment == p_alignment) {
+		return;
+	}
+	title_alignment = p_alignment;
+	queue_redraw();
+}
+
+HorizontalAlignment FoldableContainer::get_title_alignment() const {
+	return title_alignment;
+}
+
+void FoldableContainer::gui_input(const Ref<InputEvent> &p_event) {
+	ERR_FAIL_COND(p_event.is_null());
+
+	Ref<InputEventMouseMotion> m = p_event;
+
+	if (m.is_valid()) {
+		if (m->get_position().y <= title_size.height) {
+			if (!is_hovering) {
+				is_hovering = true;
+				queue_redraw();
+			}
+		} else if (is_hovering) {
+			is_hovering = false;
+			queue_redraw();
+		}
+	}
+
+	if (has_focus() && p_event->is_action_pressed("ui_accept", false, true)) {
+		set_expanded(!expanded);
+		accept_event();
+		return;
+	}
+
+	Ref<InputEventMouseButton> b = p_event;
+
+	if (b.is_valid()) {
+		if (b->is_pressed() && b->get_button_index() == MouseButton::LEFT && b->get_position().y <= title_size.height) {
+			set_expanded(!expanded);
+			accept_event();
+			return;
+		}
+	}
+}
+
+void FoldableContainer::_notification(int p_what) {
+	switch (p_what) {
+		case NOTIFICATION_DRAW: {
+			RID ci = get_canvas_item();
+
+			Ref<Texture2D> icon = expanded ? theme_cache.arrow : theme_cache.arrow_collapsed;
+			Ref<StyleBox> title_style;
+			if (is_hovering) {
+				title_style = expanded ? theme_cache.title_hover_style : theme_cache.title_collapsed_hover_style;
+			} else {
+				title_style = expanded ? theme_cache.title_style : theme_cache.title_collapsed_style;
+			}
+
+			if (title_style.is_valid()) {
+				title_size = Point2(get_size().width, title_style->get_minimum_size().height);
+			}
+			if (theme_cache.title_font.is_valid()) {
+				title_size.height += theme_cache.title_font->get_height(theme_cache.title_font_size);
+			}
+			if (icon.is_valid()) {
+				if (title_style.is_valid()) {
+					title_size.height = MAX(title_size.height, title_style->get_minimum_size().height + icon->get_height());
+				} else {
+					title_size.height = MAX(title_size.height, icon->get_height());
+				}
+			}
+
+			Point2 title_pos;
+			int title_width = 0;
+			if (title_style.is_valid()) {
+				title_style->draw(ci, Rect2(Point2(), title_size));
+				set_custom_minimum_size(Size2(0, title_size.height));
```

Comment:
we need to find a way to set the custom minimum size outside the draw notification, because it prevents setting it in the editor.